### PR TITLE
3.4.0

### PR DESCRIPTION
**Pull request by Demonkratiy**
- removed old api code; 
- removed pbiviz-update method (no reason for it now); 
- removed pbiviz-new --api flag (no reason for it now);

### CODE REVIEW

File: bin/pbiviz-package.js [hunk 1]
```diff
@@ -59,9 +59,9 @@ if (!program.pbiviz && !program.resources) {
 }
 
 VisualPackage.loadVisualPackage(cwd).then((visualPackage) => {
-    if (parseFloat(visualPackage.config.apiVersion) < parseFloat('3.8.0')) {
+    if (parseFloat(visualPackage.config.apiVersion) < parseFloat('3.2.0')) {
```
Comment (uve) [inline bin/pbiviz-package.js:62]: replace it by compare-versions package

**Review by uve — CHANGES_REQUESTED** [review]

File: bin/pbiviz-new.js [hunk 2]
```diff
@@ -64,8 +63,8 @@ if (program.force) {
     ConsoleWriter.warn('Running with force flag. Existing files will be overwritten');
 }
 
-if (program.apiVersion && parseFloat(program.apiVersion) < parseFloat('3.8.0')) {
-    ConsoleWriter.error(`Visual creation failed, please use 'powerbi-visuals-api' 3.8.0 or above to build a visual.`);
+if (program.apiVersion && parseFloat(program.apiVersion) < parseFloat('3.2.0')) {
```
[the remark this answers is not in the document]
Comment (Demonkratiy) [reply bin/pbiviz-new.js]: This comparison was removed, as it is unneeded in new implementation.

File: config.json [hunk 3]
```diff
@@ -1,32 +1 @@
...
-        "certificate": "certs/PowerBICustomVisualTest_public.crt",
-        "pfx": "certs/PowerBICustomVisualTest_public.pfx",
-        "port": "8080"
-    },
-    "visualTemplates": {
-        "circlecard": "http://github.com/microsoft/powerbi-visuals-circlecard-react/archive/master.zip"
-    }
-}
\ No newline at end of file
+{"templates":{"visuals":"templates/visuals","pbiviz":"templates/pbiviz.json.template","plugin":"templates/plugin.ts.template","package":"templates/package.json.template"},"generate":{"apiVersion":"3.8.0"},"build":{"precompileFolder":".tmp/precompile","dropFolder":".tmp/drop","js":"tmp.visual.js","css":"visual.css","stats":"../../webpack.statistics.html"},"package":{"dropFolder":"dist"},"server":{"root":"webRoot","assetsRoute":"/assets","privateKey":"certs/PowerBICustomVisualTest_private.key","certificate":"certs/PowerBICustomVisualTest_public.crt","pfx":"certs/PowerBICustomVisualTest_public.pfx","port":"8080","passphrase":"8536412297622877"},"visualTemplates":{"circlecard":"http://github.com/microsoft/powerbi-visuals-circlecard-react/archive/master.zip"}}
```
[the remark this answers is not in the document]
Comment (Demonkratiy) [reply config.json]: done

File: bin/pbiviz-start.js [hunk 4]
```diff
@@ -60,9 +56,9 @@ program.parse(options);
 let cwd = process.cwd();
 let server;
 VisualPackage.loadVisualPackage(cwd).then((visualPackage) => {
-    if (parseFloat(visualPackage.config.apiVersion) < parseFloat('3.8.0')) {
+    if (parseFloat(visualPackage.config.apiVersion) < parseFloat('3.2.0')) {
```
[the remark this answers is not in the document]
Comment (Demonkratiy) [reply bin/pbiviz-start.js]: done

File: bin/pbiviz-start.js [hunk 5]
```diff
@@ -60,9 +56,9 @@ program.parse(options);
...
         ConsoleWriter.error(`Can't start the visual because of the current API is '${visualPackage.config.apiVersion}'.
-        Please use 'powerbi-visuals-api' 3.8.0 or above to build a visual.`);
+        Please use 'powerbi-visuals-api' 3.2.0 or above to build a visual.`);
```
[the remark this answers is not in the document]
Comment (Demonkratiy) [reply bin/pbiviz-start.js]: done

File: lib/VisualGenerator.js [hunk 6]
```diff
@@ -248,8 +248,8 @@ class VisualGenerator {
     static generate(targetPath, visualName, options) {
         return new Promise((resolve, reject) => {
             let buildOptions = lodashDefaults(options, defaultOptions);
-            if (!buildOptions.apiVersion || parseFloat(buildOptions.apiVersion) < parseFloat('3.8.0')) {
-                return reject(new Error(`Can not generate a visual with an API below than '3.8.0', current API is '${buildOptions.apiVersion}'.`));
+            if (!buildOptions.apiVersion || parseFloat(buildOptions.apiVersion) < parseFloat('3.2.0')) {
```
[the remark this answers is not in the document]
Comment (Demonkratiy) [reply lib/VisualGenerator.js]: done

File: lib/WebPackWrap.js [hunk 7]
```diff
@@ -110,33 +108,28 @@ class WebPackGenerator {
...
+        let apiVersionInstalled;
+        try {
+            apiVersionInstalled = (await exec('npm list powerbi-visuals-api version')).stdout.split("\n")[1].split("@")[1].trim();
+        } catch (err) {
+            ConsoleWriter.warn(`"powerbi-visuals-api" is not installed`);
+        }
+        // if the powerbi-visual-api package wasn't installed
+        // install the powerbi-visual-api, with version from apiVersion in pbiviz.json
+        // or the latest API, if apiVersion is absent in pbiviz.json
+        if (!apiVersionInstalled || (typeof this.pbiviz.apiVersion !== "undefined" && this.pbiviz.apiVersion.slice(0, -1) != apiVersionInstalled.slice(0, -1))) {
```
[the remark this answers is not in the document]
Comment (Demonkratiy) [reply lib/WebPackWrap.js]: done

File: bin/pbiviz-package.js (hunk 1 above)
Comment (Demonkratiy) [reply bin/pbiviz-package.js]: done

File: lib/WebPackWrap.js [hunk 8]
```diff
@@ -110,33 +108,28 @@ class WebPackGenerator {
...
-            pluginConfiguration.stringResourcesSchema = api.schemas.stringResources;
-            pluginConfiguration.dependenciesSchema = api.schemas.dependencies;
-        } else {
-            pluginConfiguration.schemaLocation = path.join(process.cwd(), '.api', 'v' + this.pbiviz.apiVersion);
-            pluginConfiguration.externalJS = [path.join(visualPackage.basePath, config.build.precompileFolder, "externalJS.js")];
-            pluginConfiguration.cssStyles = path.join(visualPackage.basePath, config.build.dropFolder, config.build.css);
-            pluginConfiguration.generatePlugin = false;
+        let apiVersionInstalled;
+        try {
+            apiVersionInstalled = (await exec('npm list powerbi-visuals-api version')).stdout.split("\n")[1].split("@")[1].trim();
```
[the remark this answers is not in the document]
Comment (Demonkratiy) [reply lib/WebPackWrap.js]: done

**Review by uve — APPROVED** [review]

**Pull request merged**